src/main/java/org/commcare/cases/model/StorageIndexedTreeElementModel.java

```diff
@@ -0,0 +1,151 @@
+package org.commcare.cases.model;
+
+import org.javarosa.core.model.data.IAnswerData;
+import org.javarosa.core.model.instance.TreeElement;
+import org.javarosa.core.services.storage.IMetaData;
+import org.javarosa.core.services.storage.Persistable;
+import org.javarosa.core.util.externalizable.DeserializationException;
+import org.javarosa.core.util.externalizable.ExtUtil;
+import org.javarosa.core.util.externalizable.ExtWrapList;
+import org.javarosa.core.util.externalizable.PrototypeFactory;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.List;
+import java.util.Set;
+import java.util.Vector;
+
+/**
+ * A DB model for storing TreeElements such that particular attributes and
+ * elements are indexed and queryable using the DB.
+ *
+ * Indexed attributes/elements get their own table columns, and the rest of
+ * the TreeElement is stored as a serialized blob.
+ *
+ * @author Phillip Mates (pmates@dimagi.com)
+ */
+public class StorageIndexedTreeElementModel implements Persistable, IMetaData {
+
+    private static final String STORAGE_KEY_PREFIX = "IND_FIX_";
+    private static final String DASH_ESCAPE = "\\$\\$";
+    private static final String ATTR_PREFIX = "_$";
+    private static final String ELEM_PREFIX = "_0";
+
+    private String[] metaDataFields = null;
+    private Vector<String> indices;
+    private TreeElement root;
+
+    protected int recordId = -1;
+    protected String entityId;
+
+    @SuppressWarnings("unused")
+    public StorageIndexedTreeElementModel() {
+        // for serialization
+    }
+
+    public StorageIndexedTreeElementModel(Set<String> indices, TreeElement root) {
+        this.indices = new Vector<>(indices);
+        this.root = root;
+        metaDataFields = buildMetadataFields(this.indices);
+    }
+
+    private static String[] buildMetadataFields(List<String> indices) {
+        String[] escapedIndexList = new String[indices.size()];
+        int i = 0;
+        for (String index : indices) {
+            escapedIndexList[i++] = getColFromEntry(index);
+        }
+        return escapedIndexList;
+    }
+
+    public static String getTableName(String fixtureName) {
+        String cleanedName = fixtureName.replace(":", "_").replace(".", "_").replace("-", "_");
```

Comment:
I don't know if it's possible, but it's generally better to apply this transform once, then store the result (the place to do this may come up if we start storing the full fixture), rather than do this dynamically, as if we need to change the logic for it later, we can just keep using the retained name, rather than keeping both around and being confused about possibly using multiple ways to refer to the fixture table and checking each one.